src/TestCaseDataset.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pest;
+
+use Closure;
+use Pest\Datasets\PreprocessorRepository;
+use Pest\Repositories\DatasetsRepository;
+use ReflectionException;
+use ReflectionFunction;
+use ReflectionParameter;
+use Traversable;
+
+final class TestCaseDataset
+{
+    private PreprocessorRepository $preprocessorRepository;
+
+    /**
+     * @param Closure|iterable<int|string, mixed>|string $dataset
+     * @param array<int|string, mixed>                   $parameters
+     */
+    public function __construct(
+        private Closure|iterable|string $dataset,
+        private array $parameters,
+    ) {
+        $this->preprocessorRepository = new PreprocessorRepository();
+    }
+
+    public function isNamedDataset(): bool
+    {
+        return is_string($this->dataset);
+    }
+
+    /**
+     * @return array<int|string, mixed>
+     */
+    public function resolve(): array
+    {
+        $this->resolveFromString();
```

Comment:
I think this might be more clear if each of these were a pure method and we return the new dataset each time rather than overriding the `dataset` property.

Reply:
I implemented the way you said initially, will check why I changed to this one, don't have my IDE in front of me ATM

Reply:
The main issue was in `getPreprocessParameters()` 

in that method, actual closure-dataset parameters are extracted from the list using reflection

but at the moment of calling that method, your dataset has already been resolved to an array

now I refactored to use pure methods that return a new dataset, but for actual parameters detection the method still uses the original dataset (resolving it from string, if needed)

this should work, let me know if something is not optimal or you have a better idea